src/Juvix/Compiler/Nockma/Translation/FromAsm.hs

```diff
@@ -0,0 +1,1048 @@
+  sre $ do
+    popFrom FrameStack
+    popFrom CurrentFunction
+
+testEq :: (Members '[Compiler] r) => Sem r ()
+testEq = testEqOn ValueStack
+
+testEqOn' :: (Members '[Output (Term Natural)] r) => StackId -> Sem r ()
+testEqOn' s = output (replaceOnStackN 2 s (OpEq # stackSliceAsCell s 0 1))
+
+dumpStack :: (Members '[Compiler] r) => StackId -> Sem r ()
+dumpStack t = traceTerm (OpAddress # stackPath t)
+
+traceTerm' :: (Members '[Output (Term Natural)] r) => Term Natural -> Sem r ()
+traceTerm' t =
+  let iden = OpAddress # ([] :: Path)
+   in output (OpTrace # t # iden)
+
+incrementOn' :: (Members '[Output (Term Natural)] r) => StackId -> Sem r ()
+incrementOn' s = output (replaceOnStack s (OpInc # stackSliceAsCell s 0 0))
+
+callStdlib :: (Members '[Compiler] r) => StdlibFunction -> Sem r ()
+callStdlib = callStdlibOn ValueStack
+
+callStdlibOn' :: (Members '[Output (Term Natural)] r) => StackId -> StdlibFunction -> Sem r ()
+callStdlibOn' s f = do
+  let fNumArgs = stdlibNumArgs f
+      fPath = stdlibPath f
+      decodeFn = OpCall # (fPath # (OpAddress # stackPath StandardLibrary))
+      arguments = OpSequence # (OpAddress # [R]) # stdlibStackTake s fNumArgs
+      extractResult = (OpAddress # [L]) # (OpAddress # [R, R])
+      callFn = OpPush # (OpCall # [L] # (OpReplace # ([R, L] # arguments) # (OpAddress # [L]))) # extractResult
+
+  output (OpPush # decodeFn # callFn)
+  output (replaceTopStackN fNumArgs s)
+  where
+    stdlibStackTake :: StackId -> Natural -> Term Natural
+    stdlibStackTake sn n =
+      foldTerms
+        ( nonEmpty'
+            (take (fromIntegral n) [OpAddress # indexInStack sn i | i <- [0 ..]])
+        )
+
+save' ::
+  (Functor f, Members '[Output (Term Natural), Reader CompilerCtx] r) =>
+  Bool ->
+  m () ->
+  Sem (WithTactics Compiler f m r) (f ())
+save' isTail m = do
+  pushOntoH TempStack (OpAddress # topOfStack ValueStack)
+  popFromH ValueStack
+  runT m >>= raise . execCompilerList >>= mapM_ output >>= pureT
+  if
+      | isTail -> pureT ()
+      | otherwise -> popFromH TempStack
+
+builtinTagToTerm :: Asm.BuiltinDataTag -> Term Natural
+builtinTagToTerm = \case
+  Asm.TagTrue -> nockBoolLiteral True
+  Asm.TagFalse -> nockBoolLiteral False
+  Asm.TagReturn -> impossible
+  Asm.TagBind -> impossible
+  Asm.TagWrite -> impossible
+  Asm.TagReadLn -> impossible
+
+constructorTagToTerm :: Asm.Tag -> Term Natural
+constructorTagToTerm = \case
+  Asm.UserTag t -> OpQuote # toNock (fromIntegral (t ^. Asm.tagUserWord) :: Natural)
+  Asm.BuiltinTag b -> builtinTagToTerm b
+
+caseCmd ::
+  (Members '[Compiler] r) =>
+  Maybe (Sem r ()) ->
+  [(Asm.Tag, Sem r ())] ->
+  Sem r ()
+caseCmd defaultBranch = \case
+  [] -> sequence_ defaultBranch
+  (tag, b) : bs -> do
+    -- push the constructor tag at the top
+    push (OpAddress # topOfStack ValueStack ++ constructorPath ConstructorTag)
+    push (constructorTagToTerm tag)
+    testEq
+    branch b (caseCmd defaultBranch bs)
```

Comment:
Similarly with this branch. It doesn't need to take the value from the stack, because we know what it's branching on -- we can just do something like `[If [= [@ S] [K TagValue]] B1 B2]`.